server/recurring/cluster.go

```diff
@@ -0,0 +1,335 @@
+		p.documents[i] = document
+	}
+}
+
+func (p *PreProcessor) GetDatums() []Datum {
+	datums := make([]Datum, 0, len(p.documents))
+	for _, document := range p.documents {
+		if !document.Valid {
+			continue
+		}
+		datums = append(datums, Datum{
+			ID:          document.ID,
+			Transaction: document.Transaction,
+			String:      document.String,
+			Amount:      document.Transaction.Amount,
+			Vector:      document.Vector,
+		})
+	}
+
+	return datums
+}
+
+type Datum struct {
+	ID          uint64
+	Transaction *models.Transaction
+	String      string
+	Amount      int64
+	Vector      []float64
+}
+
+type Cluster struct {
+	ID    uint64
+	Items map[int]uint8
+}
+
+type DBSCAN struct {
+	labels    map[uint64]bool
+	dataset   []Datum
+	epsilon   float64
+	minPoints int
+	clusters  []Cluster
+}
+
+func NewDBSCAN(dataset []Datum, epsilon float64, minPoints int) *DBSCAN {
+	return &DBSCAN{
+		labels:    map[uint64]bool{},
+		dataset:   dataset,
+		epsilon:   epsilon,
+		minPoints: minPoints,
+		clusters:  nil,
+	}
+}
+
+func (d *DBSCAN) GetDatumByIndex(index int) (*Datum, bool) {
+	if index >= len(d.dataset) || index < 0 {
+		return nil, false
+	}
+
+	return &d.dataset[index], true
+}
+
+func (d *DBSCAN) Calculate() []Cluster {
+	// Initialize or reinitialize the clusters. We want to start with a clean slate.
+	d.clusters = make([]Cluster, 0)
+	// From the top, take one point at a time.
+	for index, point := range d.dataset {
+		// If we have already visited this point then skip it
+		if _, visited := d.labels[point.ID]; visited {
+			continue
+		}
+
+		// Find all the other points that are within the epsilon of this point.
+		neighbors := d.getNeighbors(index)
+		// If there are not enough points then this is not a core point.
+		if len(neighbors) < d.minPoints {
+			// Mark it as noise and keep moving
+			d.labels[point.ID] = true
+			continue
+		}
+		// Otherwise mark the point as visited so we don't do the same work again
+		d.labels[point.ID] = false
+		// Then start constructing a cluster around this point.
+		newCluster := d.expandCluster(index, neighbors)
+		// Set the cluster's unique ID to the lowest numeric ID in that cluster.
+		// HACK: I need a way to uniquely identify each cluster. Generally by using the contents of that cluster. This
+		// relies on the contents of that cluster remaining consistent over time. While the order of the clusters might
+		// change in the future or they might expand as new transactions show up, I need to know which cluster they get
+		// added to in order to tune things over time. This has the potential to cause issues on its own, what if the
+		// cluster algorithm changes enough that the "lowest ID" gets kicked out of the cluster? What if we push a bad
+		// change and the clusters change entirely? Or what if that "lowest ID" gets moved to a different cluster. This
+		// needs improvement, but I think this should be fine for the initial implementation of the clustering algorithm.
```

Comment:
I think the best way would instead be to identify each cluster by the highest value word in that cluster.